### PR TITLE
Adaptation of SRVD object handler 

### PR DESCRIPTION
**Pull request by ThomasPloski**
#3349

### CODE REVIEW

**Review by christianguenter2 — CHANGES_REQUESTED** [review]
Did not test, I guess we need to downport a few things e.g.
`if_wb_adt_plugin_resource_co`
`cl_blue_wb_utility`
`cl_srvd_wb_object_data`
and others

**Comment by mbtools** [comment]
I checked. Unfortunately, the following do **not** exist in 702:
```abap
cl_blue_wb_utility
cl_srvd_wb_object_data
cl_wb_object_operator_factory
cx_wb_object_operation_error
if_srvd_types
if_wb_adt_plugin_resource_co
if_wb_object_data_selection_co
if_wb_object_operator
````

These references exist in 702:
```abap
if_wb_object_data_model
if_wb_object_persist
```

**Comment by larshp** [comment]
there, this should help remind us, few of the error messages needs adjustments...

**Comment by ThomasPloski** [comment]
Hi Folks, Thanks for your analysis. I will do the needed changes and keep you informed.

**Comment by ThomasPloski** [comment]
Hi folks, I have resolved all points regarding non-existing objects in 702. Let's review :-)

File: src/objects/zcl_abapgit_object_srvd.clas.abap [hunk 1]
```diff
@@ -13,108 +13,90 @@ CLASS zcl_abapgit_object_srvd DEFINITION PUBLIC INHERITING FROM zcl_abapgit_obje
...
+      FOR zif_abapgit_object~is_locked .
+    ALIASES jump
+      FOR zif_abapgit_object~jump .
+    ALIASES serialize
+      FOR zif_abapgit_object~serialize .
+
+    DATA mi_persistence TYPE REF TO if_wb_object_persist .
+    DATA mv_service_definition_key TYPE seu_objkey .
+    DATA mr_service_definition TYPE REF TO data .
+    CONSTANTS mc_source_file TYPE string VALUE 'srvdsrv' ##NO_TEXT.
```
Comment (mbtools) [inline src/objects/zcl_abapgit_object_srvd.clas.abap:46]: any reason why it changed from `assrvd` (which was in line with `asddls` and `asdcls`) to this?

File: src/objects/zcl_abapgit_object_srvd.clas.abap [hunk 2]
```diff
@@ -280,48 +345,306 @@ CLASS zcl_abapgit_object_srvd IMPLEMENTATION.
...
+        iv_name = mc_xml_parent_name
+      CHANGING
+        cg_data = <ls_metadata> ).
+
+    <lv_metadata_node> = <ls_metadata>.
+
+    ASSIGN COMPONENT 'CONTENT-SOURCE' OF STRUCTURE <lg_data> TO <lv_source>.
+    ASSERT sy-subrc = 0.
+
+    <lv_source> = mo_files->read_string( 'srvdsrv' ).
```
Comment (mbtools) [inline src/objects/zcl_abapgit_object_srvd.clas.abap:556]: mc_source_file

**Comment by mbtools** [comment]
It compiles without issues on 702. I can't test features since my system doesn't support srvd.

https://github.com/abapGit-tests/SRVD will need an update

File: src/objects/zcl_abapgit_object_srvd.clas.abap (hunk 1 above)
Comment (ThomasPloski) [reply src/objects/zcl_abapgit_object_srvd.clas.abap:46]: The reason is quite simple: I have downported the already existing SRVD object handler based in SAP Cloud Platform ABAP Environment (aka Steampunk) -- see https://github.com/SAP-samples/abap-platform-refscen-flight/tree/Cloud-Platform/src/managed

File: src/objects/zcl_abapgit_object_srvd.clas.abap (hunk 2 above)
Comment (ThomasPloski) [reply src/objects/zcl_abapgit_object_srvd.clas.abap:556]: will change and provide pull requeast in some minutes

**Comment by larshp** [comment]
@christianguenter2 can you help reviewing?

**Review by christianguenter2 — CHANGES_REQUESTED** [review]
@ThomasPloski 
I updated the test repo here:
https://github.com/abapGit-tests/SRVD

When I install the test repo on S4/HANA FPS00 the service definition appears non existing after deserialization.
image
In fact it's created in the system but inactive. If I activate it manually everything looks good.
image

Seems that 
```
    zcl_abapgit_objects_activation=>add_item( ms_item ).
```
is not enough to activate service definitions properly. Old version activated service definitions AFAIR.

**Comment by christianguenter2** [comment]
@ThomasPloski 
```mi_persistence``` is not needed anymore and can be removed

**Comment by ThomasPloski** [comment]
@christianguenter2, thank you for your feedback. I will check and provide a PR update

**Comment by ThomasPloski** [comment]
@christianguenter2, please check the provided changes.

**Comment by larshp** [comment]
@christianguenter2

**Comment by mbtools** [comment]
I tested with https://github.com/abapGit-tests/SRVD on 753 SP 5.

1. Activation fails > needs to be switched from DDIC to ABAP (see comments)
2. There are some diffs > not sure if this is related to my basis release or something else

image

**Review by mbtools — CHANGES_REQUESTED** [review]

**Comment by jfilak** [comment]
Nice pull request. Any progress?

By the way, like the idea of using ADT API, however I was hoping using ADT API will need less coding.

File: src/objects/zcl_abapgit_object_srvd.clas.abap [hunk 3]
```diff
@@ -280,48 +343,306 @@ CLASS zcl_abapgit_object_srvd IMPLEMENTATION.
...
+  ENDMETHOD.
+
+
+  METHOD is_delete_tadir.
+    rv_delete_tadir = abap_true.
+  ENDMETHOD.
+
+
+  METHOD is_ddic.
+    rv_ddic = abap_true.
```
[the remark this answers is not in the document]
Comment (ThomasPloski) [reply src/objects/zcl_abapgit_object_srvd.clas.abap]: PR provided

File: src/objects/zcl_abapgit_object_srvd.clas.abap [hunk 4]
```diff
@@ -237,10 +301,9 @@ CLASS zcl_abapgit_object_srvd IMPLEMENTATION.
```
[the remark this answers is not in the document]
Comment (ThomasPloski) [reply src/objects/zcl_abapgit_object_srvd.clas.abap:301]: PR provided

**Comment by ThomasPloski** [comment]
> I tested with https://github.com/abapGit-tests/SRVD on 753 SP 5.
> 
> 1. Activation fails > needs to be switched from DDIC to ABAP (see comments)
> 2. There are some diffs > not sure if this is related to my basis release or something else
> 
> image

This results because the exchange file format has been changed to that one which is used in Steampunk. Please create a new SRVD and export it. Afterwards make some changes directly in the repo and try to import it.

**Comment by ThomasPloski** [comment]
@mbtools I have adjusted your requested changes. Please have a look and verify if the behaviour fits :-)

File: src/objects/zcl_abapgit_object_srvd.clas.abap [hunk 5]
```diff
@@ -280,48 +343,306 @@ CLASS zcl_abapgit_object_srvd IMPLEMENTATION.
...
+        iv_name = mc_xml_parent_name
+      CHANGING
+        cg_data = <ls_metadata> ).
+
+    <lv_metadata_node> = <ls_metadata>.
+
+    ASSIGN COMPONENT 'CONTENT-SOURCE' OF STRUCTURE <lg_data> TO <lv_source>.
+    ASSERT sy-subrc = 0.
+
+    <lv_source> = mo_files->read_string( mc_source_file ).
```
Comment (mbtools) [inline src/objects/zcl_abapgit_object_srvd.clas.abap:554]: the change is not compatible to existing repos which use `assrvd` extension. maybe add this

```abap
if <lv_source> is initial.
    <lv_source> = mo_files->read_string( 'assrvd' ).
endif.
```

**Review by mbtools — COMMENTED** [review]
aside from the compatibility, it looks ok to me now

File: src/objects/zcl_abapgit_object_srvd.clas.abap (hunk 5 above)
Comment (ThomasPloski) [reply src/objects/zcl_abapgit_object_srvd.clas.abap:554]: In general, you are right. But currently there are two different exchange formats in place. With your suggested coding change you will be able to import SRVD with assrvd format. But the same question will come up regarding export. My suggestion would be to rename existing files to SAP provided format. I will provide some demo data.

**Comment by ThomasPloski** [comment]
@mbtools, I have put [here](https://github.com/abapGit-tests/SRVD_w_DDLS.git) the example

File: src/objects/zcl_abapgit_object_srvd.clas.abap (hunk 5 above)
Comment (ThomasPloski) [reply src/objects/zcl_abapgit_object_srvd.clas.abap:554]: @mbtools, I have add the if-endif-part to branch. Please verify and merge.

**Review by mbtools — APPROVED** [review]

File: src/objects/zcl_abapgit_object_srvd.clas.abap (hunk 5 above)
Comment (mbtools) [reply src/objects/zcl_abapgit_object_srvd.clas.abap:554]: yes, it's one way. but we will be able to pull the old and stage new format. thanks

**Comment by mbtools** [comment]
@larshp lgtm

**Comment by larshp** [comment]
👍 planning to merge this tomorrow

**Comment by larshp** [comment]
waiting until ~sunday, #dontShipOnFriday

**Pull request merged**